### PR TITLE
JEP draft: WebSocket Services

### PR DESCRIPTION
**Pull request by jglick**
Prototype implementation in https://github.com/jenkinsci/jenkins/pull/4369 and its upstreams.

### CODE REVIEW

File: jep/0000/README.adoc [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+
+== Abstract
+
+Jenkins offers an API allowing WebSocket services to be implemented.
+This is available using the Stapler `HttpResponse` idiom,
+served as the final action after the regular Stapler routing mechanism.
+Unlike regular HTTP endpoints, bidirectional streaming communication is possible.
+
+The principal service exposed in this way is a new Remoting transport for inbound build agents at an HTTP(S) URL.
+The agent itself (typically named `slave.jar`) includes a WebSocket client capable of connecting to this endpoint.
```
Comment (slide) [inline jep/0000/README.adoc]: I think the Jenkins project is trying to move people to use `agent.jar` instead of `slave.jar`.

Comment (jglick) [reply jep/0000/README.adoc]: Yes, the current launch script in the Docker image reflects the rename. I will update this accordingly.

File: jep/0000/README.adoc [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+
+The principal service exposed in this way is a new Remoting transport for inbound build agents at an HTTP(S) URL.
+The agent itself (typically named `agent.jar`) includes a WebSocket client capable of connecting to this endpoint.
+This provides an alternative to using the `JNLP4-connect` protocol on the Jenkins TCP port.
+Otherwise the regular behavior of inbound agents (`JNLPLauncher`) applies.
+
+The Jenkins CLI could also be served via WebSocket in place of the SSHD port or the problematic “full-duplex HTTP” transport.
+
+The implementation relies on APIs avalable in the built-in Jetty (“Winstone”) servlet container,
+but Jetty APIs are not reëxported to plugins.
```
Comment (jeffret-b) [inline jep/0000/README.adoc]:
```suggestion
but Jetty APIs are not re-exported to plugins.
```

Comment (jglick) [reply jep/0000/README.adoc]: Either is valid; I am using the older notation.

**Comment by jglick** [comment]
@rtyler I guess this can be merged & numbered now?

**Comment by oleg-nenashev** [comment]
I might get to merging it next week after DWJW

File: jep/0000/README.adoc (hunk 2 above)
Comment (jeffret-b) [reply jep/0000/README.adoc]: True, but the newer form would be more recognizable and readable to more people.

**Comment by bitwiseman** [comment]
@oleg-nenashev  No reason to wait.

**Comment by oleg-nenashev** [comment]
No reason, just a bandwidth issue

**Pull request merged**